content/blog/posts/2022-08-04-add-desktop-notifications.md

```diff
@@ -0,0 +1,241 @@
+
+This gets us a basic notification, but without any interaction from the user. To do that, we have to add actions to the options hash that you can watch and respond too. The spec requires that we only can do that if we're triggering the notification via a service worker and responding to events there. Let's create a basic service worker file and add a listener. Create a file called `service-worker.js` and add the following code.
+
+```js
+self.addEventListener('notificationclick', function(event) {
+  const clickedNotification = event.notification;
+  clickedNotification.close();
+
+  // Do something as the result of the notification click
+  const promiseChain = doSomething();
+  event.waitUntil(promiseChain);
+});
+```
+
+In this we're watching for a `notificationclick` event and want to do something with that once we know the user has interacted with it. We'll also want to register our worker in the notification service and have it trigger the notification flow.
+
+```js
+_startServiceWorker(notificationService) {
+  if ('serviceWorker' in navigator) {
+    navigator.serviceWorker.register('service-worker.js')
+    .then(function(registration) {
+      registration.addEventListener('updatefound', function() {
+        // If updatefound is fired, it means that there's
+        // a new service worker being installed.
+        const installingWorker = registration.installing;
+        console.log('A new service worker is being installed:',
+          installingWorker);
+
+        // You can listen for changes to the installing service worker's
+        // state via installingWorker.onstatechange
+      });
+      
+      notificationService._showNotification(registration);
+    })
+    .catch(function(error) {
+      console.log('Service worker registration failed:', error);
+    });
+  } else {
+    console.log('Service workers are not supported.');
+  }
+}
+```
+
+In the above, we check for service worker support and then register our newly created worker file. After it's installled we show the notification in an updated `_showNotification` method. 
+
+```js
+async _showNotification(worker) {
+  const permission = await this._isShowNotificationPermitted();
+  if (permission !== 'denied') {
+    worker.showNotification('Be notified of new posts', {
+      body: 'Let us notifiy you of new posts on our site!',
+      requireInteraction: true,
+      actions: [
+        {
+          action: 'dismiss',
+          title: 'Dismiss'
+        },
+        {
+          action: 'subscribe',
+          title: 'Subscribe'
+        }
+      ]
+    });
+  }
+}
+```
+
+The difference now is that we can set actions on the notification options and allow the user to select something from that notification. This click will get passed to the worker and given what the action is, we can react and do something. So, let's do just that and act on the user's response, which we're requiring with `requireInteraction`.
```

Comment:
```suggestion
These changes allow us to set actions on the notification options and allow the user to interact with the notification, in this case by clicking [what they're clicking]. This click will get passed to the worker so we can react with a response. Let's do just that and act on the user's response, with `requireInteraction`.
```

- maybe show the before and after together so the change is more visible/memorable
- not sure if the suggested change is true to your original meaning